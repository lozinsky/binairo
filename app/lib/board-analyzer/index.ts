export * from './analyze-board-by-kind';
export * from './analyze-board';
