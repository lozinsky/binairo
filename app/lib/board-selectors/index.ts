export * from './select-filled-imbalanced-line';
export * from './select-line-that-can-be-equal-to-another';
export * from './select-line-that-includes-another';
export * from './select-line-that-is-equal-to-others';
export * from './select-middle-cell-between-two-identical';
export * from './select-next-cell-after-two-identical';
export * from './select-three-or-more-identical-sequential-cells';
export * from './select-unfilled-balanced-line';
