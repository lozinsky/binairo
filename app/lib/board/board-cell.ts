export type BoardCellKind = 'fixed' | 'regular';

export type BoardCellKindRaw = 0 | 1;

export type BoardCellState = 'B' | 'E' | 'R';

export type BoardCellStateRaw = 0 | 1 | 2;

export function isBoardCellKindRaw(value: unknown): value is BoardCellKindRaw {
  return typeof value === 'number' && value in BOARD_CELL_KIND_BY_RAW;
}

export function isBoardCellStateRaw(value: unknown): value is BoardCellStateRaw {
  return typeof value === 'number' && value in BOARD_CELL_STATE_BY_RAW;
}

const NEXT_BOARD_CELL_STATE_BY_PREV_BOARD_CELL_STATE: Readonly<Record<BoardCellState, BoardCellState>> = {
  B: 'E',
  E: 'R',
  R: 'B',
};

const BALANCE_BY_BOARD_CELL_STATE: Readonly<Record<BoardCellState, number>> = {
  B: -1,
  E: 0,
  R: 1,
};

const RAW_BY_BOARD_CELL_KIND: Readonly<Record<BoardCellKind, BoardCellKindRaw>> = {
  fixed: 0,
  regular: 1,
};

const BOARD_CELL_KIND_BY_RAW: Readonly<Record<BoardCellKindRaw, BoardCellKind>> = {
  0: 'fixed',
  1: 'regular',
};

const RAW_BY_BOARD_CELL_STATE: Readonly<Record<BoardCellState, BoardCellStateRaw>> = {
  B: 0,
  E: 2,
  R: 1,
};

const BOARD_CELL_STATE_BY_RAW: Readonly<Record<BoardCellStateRaw, BoardCellState>> = {
  0: 'B',
  1: 'R',
  2: 'E',
};

export type BoardCellValue = readonly [kind: BoardCellKindRaw, state: BoardCellStateRaw];

export class BoardCell {
  get balance() {
    return BALANCE_BY_BOARD_CELL_STATE[this.#state];
  }

  get isEmpty() {
    return this.#state === 'E';
  }

  get isFilled() {
    return !this.isEmpty;
  }

  get isFixed() {
    return this.#kind === 'fixed';
  }

  get state() {
    return this.#state;
  }

  readonly #kind: BoardCellKind;
  readonly #state: BoardCellState;

  constructor(kind: BoardCellKind, state: BoardCellState) {
    this.#kind = kind;
    this.#state = state;
  }

  static create(state: BoardCellState) {
    return new this('regular', state);
  }

  static from(value: BoardCellValue) {
    return new this(BOARD_CELL_KIND_BY_RAW[value[0]], BOARD_CELL_STATE_BY_RAW[value[1]]);
  }

  equals(other: BoardCell) {
    return this.#state === other.#state;
  }

  next() {
    return this.isFixed ? this : new BoardCell(this.#kind, NEXT_BOARD_CELL_STATE_BY_PREV_BOARD_CELL_STATE[this.#state]);
  }

  toFixed() {
    return new BoardCell('fixed', this.#state);
  }

  valueOf(): BoardCellValue {
    return [RAW_BY_BOARD_CELL_KIND[this.#kind], RAW_BY_BOARD_CELL_STATE[this.#state]];
  }
}

export function isBoardCellValue(value: unknown): value is BoardCellValue {
  if (Array.isArray(value) && value.length === 2) {
    return isBoardCellKindRaw(value[0]) && isBoardCellStateRaw(value[1]);
  }

  return false;
}
