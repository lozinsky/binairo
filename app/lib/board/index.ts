export * from './board-cell';
export * from './board-line';
export * from './board';
