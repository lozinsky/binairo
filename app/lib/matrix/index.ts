export * from './matrix-line';
export * from './matrix-rotation';
export * from './matrix-selection';
export * from './matrix';
