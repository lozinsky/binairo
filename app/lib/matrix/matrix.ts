import type { MatrixLine, MatrixLineCell } from './matrix-line';

export interface Matrix<L extends MatrixLine<unknown>> {
  at(index: number): L | undefined;
  entries(): Iterable<[number, L]>;
}

export type MatrixCell<M> = M extends Matrix<infer L> ? MatrixLineCell<L> : never;
