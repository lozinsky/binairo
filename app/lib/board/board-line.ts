import { DoubleMatrixRotation, type MatrixLine, type MatrixRotatable } from '~/lib/matrix';
import { expectToBeDefined } from '~/shared/expect';

import { BoardCell, BoardCellState, type BoardCellValue } from './board-cell';

export type BoardLineValue = readonly BoardCellValue[];

export class BoardLine implements Iterable<BoardCell>, MatrixLine<BoardCell>, MatrixRotatable<BoardLine> {
  readonly #cells: readonly BoardCell[];

  constructor(cells: readonly BoardCell[]) {
    this.#cells = cells;
  }

  static blank(size: number) {
    return new this(Array.from({ length: size }, () => BoardCell.create(BoardCellState.E)));
  }

  static create(states: readonly BoardCellState[]) {
    return new this(states.map((state) => BoardCell.create(state)));
  }

  static from(value: BoardLineValue) {
    return new this(value.map((value) => BoardCell.from(value)));
  }

  [Symbol.iterator]() {
    return this.#cells[Symbol.iterator]();
  }

  at(index: number) {
    return this.#cells.at(index);
  }

  entries() {
    return this.#cells.entries();
  }

  equals(other: BoardLine) {
    if (this.length !== other.length) {
      return false;
    }

    return this.#cells.every((cell, index) => cell.equals(expectToBeDefined(other.at(index))));
  }

  includes(other: BoardLine) {
    if (this.length !== other.length) {
      return false;
    }

    return this.#cells.every((a, index) => {
      const b = expectToBeDefined(other.at(index));

      return b.isEmpty ? true : a.equals(b);
    });
  }

  indexOf(cell: BoardCell) {
    return this.#cells.indexOf(cell);
  }

  keys() {
    return this.#cells.keys();
  }

  lengthOf(state: BoardCellState) {
    return this.#cells.reduce((length, cell) => (cell.state === state ? length + 1 : length), 0);
  }

  rotate() {
    return new BoardLine(this.#cells.toReversed());
  }

  slice(start: number, end?: number) {
    return this.#cells.slice(start, end);
  }

  toDoubleRotation() {
    return DoubleMatrixRotation.run<BoardLine>(this);
  }

  valueOf(): BoardLineValue {
    return this.#cells.map((cell) => cell.valueOf());
  }

  get balance() {
    return this.#cells.reduce((balance, cell) => balance + cell.balance, 0);
  }

  get isEmpty() {
    return this.#cells.every((cell) => cell.isEmpty);
  }

  get isFilled() {
    return this.#cells.every((cell) => cell.isFilled);
  }

  get length() {
    return this.#cells.length;
  }

  get progress() {
    return this.#cells.reduce((progress, cell) => (cell.isFilled ? progress + 1 : progress), 0) / this.length;
  }
}
