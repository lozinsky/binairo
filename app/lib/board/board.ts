import {
  DoubleMatrixRotation,
  type Matrix,
  type MatrixReversible,
  type MatrixRotatable,
  type MatrixSelection,
  QuadrupleMatrixRotation,
} from '~/lib/matrix';
import { assert } from '~/shared/assert';
import { expectToBeDefined } from '~/shared/expect';

import type { BoardCell, BoardCellState } from './board-cell';

import { BoardLine, type BoardLineValue, isBoardLineValue } from './board-line';

export type BoardOrientation = 'landscape' | 'portrait';

const NEXT_BOARD_ORIENTATION_BY_PREV_BOARD_ORIENTATION: Readonly<Record<BoardOrientation, BoardOrientation>> = {
  landscape: 'portrait',
  portrait: 'landscape',
};

export type BoardValue = readonly BoardLineValue[];

export class Board implements Iterable<BoardLine>, Matrix<BoardLine>, MatrixReversible<Board>, MatrixRotatable<Board> {
  get length() {
    return this.#lines.length;
  }

  get orientation() {
    return this.#orientation;
  }

  get progress() {
    return this.#lines.reduce((progress, line) => progress + line.progress, 0) / this.#lines.length;
  }

  readonly #lines: readonly BoardLine[];
  readonly #orientation: BoardOrientation;

  constructor(orientation: BoardOrientation, lines: readonly BoardLine[]) {
    for (const line of lines) {
      assert(line.length === lines.length, 'Expected line length to be equal to board size');
    }

    this.#lines = lines;
    this.#orientation = orientation;
  }

  static blank(size: number) {
    return new this(
      'portrait',
      Array.from({ length: size }, () => BoardLine.blank(size)),
    );
  }

  static create(states: ReadonlyArray<readonly BoardCellState[]>) {
    return new this(
      'portrait',
      states.map((states) => BoardLine.create(states)),
    );
  }

  static from(value: BoardValue) {
    return new this(
      'portrait',
      value.map((value) => BoardLine.from(value)),
    );
  }

  at(index: number) {
    return this.#lines.at(index);
  }

  entries() {
    return this.#lines.entries();
  }

  replace(index: number, callback: (line: BoardLine) => BoardLine) {
    return new Board(this.#orientation, this.#lines.with(index, callback(expectToBeDefined(this.#lines.at(index)))));
  }

  replaceBy(selection: MatrixSelection, callback: (cell: BoardCell) => BoardCell) {
    const content = selection.execute(this);
    const lines = this.#lines.map((line) => {
      const cells = Array.from(line);

      for (const cell of content) {
        const index = line.indexOf(cell);

        if (index === -1) {
          continue;
        }

        cells[index] = callback(expectToBeDefined(cells[index]));
      }

      return new BoardLine(cells);
    });

    return new Board(this.#orientation, lines);
  }

  reverse() {
    return new Board(this.#orientation, this.#lines.toReversed());
  }

  rotate() {
    const orientation = NEXT_BOARD_ORIENTATION_BY_PREV_BOARD_ORIENTATION[this.#orientation];

    if (this.#lines.length === 0) {
      return new Board(orientation, []);
    }

    const lines: BoardLine[] = [];

    for (const index of expectToBeDefined(this.#lines[0]).keys()) {
      const cells: BoardCell[] = [];

      for (const line of this.#lines) {
        cells.push(expectToBeDefined(line.at(index)));
      }

      lines.push(new BoardLine(cells));
    }

    return new Board(orientation, lines);
  }

  slice(start: number, end?: number) {
    return this.#lines.slice(start, end);
  }

  [Symbol.iterator]() {
    return this.#lines[Symbol.iterator]();
  }

  toDoubleRotation() {
    return DoubleMatrixRotation.run<Board>(this);
  }

  toQuadrupleRotation() {
    return QuadrupleMatrixRotation.run<Board>(this);
  }

  valueOf(): BoardValue {
    return this.#lines.map((line) => line.valueOf());
  }
}

export function isBoardValue(value: unknown): value is BoardValue {
  return Array.isArray(value) && value.every(isBoardLineValue);
}
