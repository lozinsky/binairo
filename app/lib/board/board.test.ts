import { expect, test, vi } from 'vitest';

import { MatrixSelection } from '~/lib/matrix';

import type { BoardCellState } from './board-cell';
import type { BoardCell } from './board-cell';

import { Board, type BoardValue } from './board';
import { BoardLine } from './board-line';

test.each([4, 6, 8, 10])('returns blank board', (size) => {
  expect(Board.blank(size)).toMatchSnapshot();
});

test.each<[BoardCellState[][]]>([
  [
    [
      ['B', 'R', 'B', 'R'],
      ['R', 'B', 'R', 'B'],
      ['R', 'B', 'B', 'R'],
      ['B', 'R', 'R', 'B'],
    ],
  ],
  [
    [
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ],
  ],
])('returns board from states', (states) => {
  expect(Board.create(states)).toMatchSnapshot();
});

test.each([
  [
    [
      [
        [1, 0],
        [1, 1],
        [1, 0],
        [1, 1],
      ],
      [
        [1, 1],
        [1, 0],
        [1, 1],
        [1, 0],
      ],
      [
        [1, 1],
        [1, 0],
        [1, 0],
        [1, 1],
      ],
      [
        [1, 0],
        [1, 1],
        [1, 1],
        [1, 0],
      ],
    ] as BoardValue,
  ],
  [
    [
      [
        [1, 0],
        [1, 1],
        [1, 0],
      ],
      [
        [1, 1],
        [1, 0],
        [1, 1],
      ],
      [
        [1, 1],
        [1, 0],
        [1, 0],
      ],
    ] as BoardValue,
  ],
])('returns board from value of', (value) => {
  expect(Board.from(value)).toMatchSnapshot();
});

test.each([
  Board.create([
    ['B', 'R', 'B', 'R'],
    ['R', 'B', 'R', 'B'],
    ['R', 'B', 'B', 'R'],
    ['B', 'R', 'R', 'B'],
  ]),
  Board.create([
    ['B', 'R', 'B'],
    ['R', 'B', 'R'],
    ['R', 'B', 'B'],
  ]),
])('returns board iterator', (board) => {
  expect(Array.from(board)).toMatchSnapshot();
});

test.each([
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    -1,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    0,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    1,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    2,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    3,
  ],
])('returns board line at index', (board, index) => {
  expect(board.at(index)).toMatchSnapshot();
});

test.each([
  Board.create([
    ['B', 'R', 'B', 'R'],
    ['R', 'B', 'R', 'B'],
    ['R', 'B', 'B', 'R'],
    ['B', 'R', 'R', 'B'],
  ]),
  Board.create([
    ['B', 'R', 'B'],
    ['R', 'B', 'R'],
    ['R', 'B', 'B'],
  ]),
])('returns board entries', (board) => {
  expect(Array.from(board.entries())).toMatchSnapshot();
});

test.each([
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    -1,
    vi.fn(() => BoardLine.blank(3)),
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    0,
    vi.fn(() => BoardLine.blank(3)),
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    1,
    vi.fn(() => BoardLine.blank(3)),
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    2,
    vi.fn(() => BoardLine.blank(3)),
  ],
])('replaces board line at index in board', (board, index, callback) => {
  expect(board.replace(index, callback)).toMatchSnapshot();
  expect(callback).toBeCalledWith(board.at(index));
});

test.each([
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    new MatrixSelection([{ x: -1, y: -1 }]),
    vi.fn((cell: BoardCell) => cell.next()),
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    new MatrixSelection([{ x: 0, y: 0 }]),
    vi.fn((cell: BoardCell) => cell.next()),
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    new MatrixSelection([{ x: 1, y: 1 }]),
    vi.fn((cell: BoardCell) => cell.next()),
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    new MatrixSelection([
      { x: 0, y: 0 },
      { x: 0, y: 0 },
      { x: 0, y: 0 },
      { x: 1, y: 1 },
      { x: 1, y: 1 },
      { x: 1, y: 1 },
      { x: -1, y: -1 },
      { x: -1, y: -1 },
      { x: -1, y: -1 },
    ]),
    vi.fn((cell: BoardCell) => cell.next()),
  ],
])('replaces board cell at selection in board', (board, selection, callback) => {
  expect(board.replaceBy(selection, callback)).toMatchSnapshot();
});

test.each([
  Board.create([
    ['B', 'R', 'B', 'R'],
    ['R', 'B', 'R', 'B'],
    ['R', 'B', 'B', 'R'],
    ['B', 'R', 'R', 'B'],
  ]),
  Board.create([
    ['B', 'R', 'B'],
    ['R', 'B', 'R'],
    ['R', 'B', 'B'],
  ]),
])('reverses board', (board) => {
  expect(board.reverse()).toMatchSnapshot();
  expect(board.reverse().orientation).toMatchSnapshot();
});

test.each([
  Board.create([
    ['B', 'R', 'B', 'R'],
    ['R', 'B', 'R', 'B'],
    ['R', 'B', 'B', 'R'],
    ['B', 'R', 'R', 'B'],
  ]),
  Board.create([
    ['B', 'R', 'B'],
    ['R', 'B', 'R'],
    ['R', 'B', 'B'],
  ]),
  Board.create([]),
])('rotates board', (board) => {
  expect(board.rotate()).toMatchSnapshot();
  expect(board.rotate().orientation).toMatchSnapshot();
});

test.each([
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    -1,
    undefined,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    0,
    undefined,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    1,
    undefined,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    2,
    undefined,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    3,
    undefined,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    0,
    -1,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    0,
    0,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    0,
    1,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    0,
    2,
  ],
  [
    Board.create([
      ['B', 'R', 'B'],
      ['R', 'B', 'R'],
      ['R', 'B', 'B'],
    ]),
    0,
    3,
  ],
])('slices board', (board, start, end) => {
  expect(board.slice(start, end)).toMatchSnapshot();
});

test.each([
  Board.create([
    ['B', 'R', 'B', 'R'],
    ['R', 'B', 'R', 'B'],
    ['R', 'B', 'B', 'R'],
    ['B', 'R', 'R', 'B'],
  ]),
  Board.create([
    ['B', 'R', 'B'],
    ['R', 'B', 'R'],
    ['R', 'B', 'B'],
  ]),
])('returns board value of', (board) => {
  expect(board.valueOf()).toMatchSnapshot();
});

test.each([
  Board.create([
    ['B', 'R', 'B', 'R'],
    ['R', 'B', 'R', 'B'],
    ['R', 'B', 'B', 'R'],
    ['B', 'R', 'R', 'B'],
  ]),
  Board.create([
    ['B', 'R', 'B'],
    ['R', 'B', 'R'],
    ['R', 'B', 'B'],
  ]),
])('returns board length', (board) => {
  expect(board.length).toMatchSnapshot();
});

test.each([new Board('portrait', []), new Board('landscape', [])])('returns board orientation', (board) => {
  expect(board.orientation).toMatchSnapshot();
});

test.each([
  Board.create([
    ['E', 'E', 'E', 'E'],
    ['E', 'E', 'E', 'E'],
    ['E', 'E', 'E', 'E'],
    ['E', 'E', 'E', 'E'],
  ]),
  Board.create([
    ['B', 'E', 'E', 'E'],
    ['E', 'E', 'R', 'E'],
    ['E', 'E', 'E', 'E'],
    ['E', 'R', 'R', 'E'],
  ]),
  Board.create([
    ['B', 'R', 'B', 'R'],
    ['R', 'B', 'R', 'B'],
    ['R', 'B', 'B', 'R'],
    ['B', 'R', 'R', 'B'],
  ]),
])('returns board progress', (board) => {
  expect(board.progress).toMatchSnapshot();
});
