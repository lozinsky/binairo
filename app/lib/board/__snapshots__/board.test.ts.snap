// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`replaces board cell at selection in board 1`] = `
[🟦] [🟥] [🟦]
[🟥] [🟦] [🟥]
[🟥] [🟦] [⬜]
`;

exports[`replaces board cell at selection in board 2`] = `
[⬜] [🟥] [🟦]
[🟥] [🟦] [🟥]
[🟥] [🟦] [🟦]
`;

exports[`replaces board cell at selection in board 3`] = `
[🟦] [🟥] [🟦]
[🟥] [⬜] [🟥]
[🟥] [🟦] [🟦]
`;

exports[`replaces board cell at selection in board 4`] = `
[🟦] [🟥] [🟦]
[🟥] [🟦] [🟥]
[🟥] [🟦] [🟦]
`;

exports[`replaces board line at index in board 1`] = `
[🟦] [🟥] [🟦]
[🟥] [🟦] [🟥]
[⬜] [⬜] [⬜]
`;

exports[`replaces board line at index in board 2`] = `
[⬜] [⬜] [⬜]
[🟥] [🟦] [🟥]
[🟥] [🟦] [🟦]
`;

exports[`replaces board line at index in board 3`] = `
[🟦] [🟥] [🟦]
[⬜] [⬜] [⬜]
[🟥] [🟦] [🟦]
`;

exports[`replaces board line at index in board 4`] = `
[🟦] [🟥] [🟦]
[🟥] [🟦] [🟥]
[⬜] [⬜] [⬜]
`;

exports[`returns blank board 1`] = `
[⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜]
`;

exports[`returns blank board 2`] = `
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
`;

exports[`returns blank board 3`] = `
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
`;

exports[`returns blank board 4`] = `
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]
`;

exports[`returns board entries 1`] = `
[
  [
    0,
    [🟦] [🟥] [🟦] [🟥],
  ],
  [
    1,
    [🟥] [🟦] [🟥] [🟦],
  ],
  [
    2,
    [🟥] [🟦] [🟦] [🟥],
  ],
  [
    3,
    [🟦] [🟥] [🟥] [🟦],
  ],
]
`;

exports[`returns board entries 2`] = `
[
  [
    0,
    [🟦] [🟥] [🟦],
  ],
  [
    1,
    [🟥] [🟦] [🟥],
  ],
  [
    2,
    [🟥] [🟦] [🟦],
  ],
]
`;

exports[`returns board from states 1`] = `
[🟦] [🟥] [🟦] [🟥]
[🟥] [🟦] [🟥] [🟦]
[🟥] [🟦] [🟦] [🟥]
[🟦] [🟥] [🟥] [🟦]
`;

exports[`returns board from states 2`] = `
[🟦] [🟥] [🟦]
[🟥] [🟦] [🟥]
[🟥] [🟦] [🟦]
`;

exports[`returns board from string 1`] = `
[🟦] [🟥] [🟦] [🟥]
[🟥] [🟦] [🟥] [🟦]
[🟥] [🟦] [🟦] [🟥]
[🟦] [🟥] [🟥] [🟦]
`;

exports[`returns board from string 2`] = `
[🟦] [🟥] [🟦]
[🟥] [🟦] [🟥]
[🟥] [🟦] [🟦]
`;

exports[`returns board from value of 1`] = `
[🟦] [🟥] [🟦] [🟥]
[🟥] [🟦] [🟥] [🟦]
[🟥] [🟦] [🟦] [🟥]
[🟦] [🟥] [🟥] [🟦]
`;

exports[`returns board from value of 2`] = `
[🟦] [🟥] [🟦]
[🟥] [🟦] [🟥]
[🟥] [🟦] [🟦]
`;

exports[`returns board iterator 1`] = `
[
  [🟦] [🟥] [🟦] [🟥],
  [🟥] [🟦] [🟥] [🟦],
  [🟥] [🟦] [🟦] [🟥],
  [🟦] [🟥] [🟥] [🟦],
]
`;

exports[`returns board iterator 2`] = `
[
  [🟦] [🟥] [🟦],
  [🟥] [🟦] [🟥],
  [🟥] [🟦] [🟦],
]
`;

exports[`returns board length 1`] = `4`;

exports[`returns board length 2`] = `3`;

exports[`returns board line at index 1`] = `[🟥] [🟦] [🟦]`;

exports[`returns board line at index 2`] = `[🟦] [🟥] [🟦]`;

exports[`returns board line at index 3`] = `[🟥] [🟦] [🟥]`;

exports[`returns board line at index 4`] = `[🟥] [🟦] [🟦]`;

exports[`returns board line at index 5`] = `undefined`;

exports[`returns board orientation 1`] = `1`;

exports[`returns board orientation 2`] = `0`;

exports[`returns board progress 1`] = `0`;

exports[`returns board progress 2`] = `0.25`;

exports[`returns board progress 3`] = `1`;

exports[`returns board string 1`] = `"W1tbMSwwXSxbMSwxXSxbMSwwXSxbMSwxXV0sW1sxLDFdLFsxLDBdLFsxLDFdLFsxLDBdXSxbWzEsMV0sWzEsMF0sWzEsMF0sWzEsMV1dLFtbMSwwXSxbMSwxXSxbMSwxXSxbMSwwXV1d"`;

exports[`returns board string 2`] = `"W1tbMSwwXSxbMSwxXSxbMSwwXV0sW1sxLDFdLFsxLDBdLFsxLDFdXSxbWzEsMV0sWzEsMF0sWzEsMF1dXQ"`;

exports[`returns board value of 1`] = `
[
  [
    [
      1,
      0,
    ],
    [
      1,
      1,
    ],
    [
      1,
      0,
    ],
    [
      1,
      1,
    ],
  ],
  [
    [
      1,
      1,
    ],
    [
      1,
      0,
    ],
    [
      1,
      1,
    ],
    [
      1,
      0,
    ],
  ],
  [
    [
      1,
      1,
    ],
    [
      1,
      0,
    ],
    [
      1,
      0,
    ],
    [
      1,
      1,
    ],
  ],
  [
    [
      1,
      0,
    ],
    [
      1,
      1,
    ],
    [
      1,
      1,
    ],
    [
      1,
      0,
    ],
  ],
]
`;

exports[`returns board value of 2`] = `
[
  [
    [
      1,
      0,
    ],
    [
      1,
      1,
    ],
    [
      1,
      0,
    ],
  ],
  [
    [
      1,
      1,
    ],
    [
      1,
      0,
    ],
    [
      1,
      1,
    ],
  ],
  [
    [
      1,
      1,
    ],
    [
      1,
      0,
    ],
    [
      1,
      0,
    ],
  ],
]
`;

exports[`reverses board 1`] = `
[🟦] [🟥] [🟥] [🟦]
[🟥] [🟦] [🟦] [🟥]
[🟥] [🟦] [🟥] [🟦]
[🟦] [🟥] [🟦] [🟥]
`;

exports[`reverses board 2`] = `1`;

exports[`reverses board 3`] = `
[🟥] [🟦] [🟦]
[🟥] [🟦] [🟥]
[🟦] [🟥] [🟦]
`;

exports[`reverses board 4`] = `1`;

exports[`rotates board 1`] = `
[🟦] [🟥] [🟥] [🟦]
[🟥] [🟦] [🟦] [🟥]
[🟦] [🟥] [🟦] [🟥]
[🟥] [🟦] [🟥] [🟦]
`;

exports[`rotates board 2`] = `0`;

exports[`rotates board 3`] = `
[🟦] [🟥] [🟥]
[🟥] [🟦] [🟦]
[🟦] [🟥] [🟦]
`;

exports[`rotates board 4`] = `0`;

exports[`rotates board 5`] = ``;

exports[`rotates board 6`] = `0`;

exports[`slices board 1`] = `
[
  [🟥] [🟦] [🟦],
]
`;

exports[`slices board 2`] = `
[
  [🟦] [🟥] [🟦],
  [🟥] [🟦] [🟥],
  [🟥] [🟦] [🟦],
]
`;

exports[`slices board 3`] = `
[
  [🟥] [🟦] [🟥],
  [🟥] [🟦] [🟦],
]
`;

exports[`slices board 4`] = `
[
  [🟥] [🟦] [🟦],
]
`;

exports[`slices board 5`] = `[]`;

exports[`slices board 6`] = `
[
  [🟦] [🟥] [🟦],
  [🟥] [🟦] [🟥],
]
`;

exports[`slices board 7`] = `[]`;

exports[`slices board 8`] = `
[
  [🟦] [🟥] [🟦],
]
`;

exports[`slices board 9`] = `
[
  [🟦] [🟥] [🟦],
  [🟥] [🟦] [🟥],
]
`;

exports[`slices board 10`] = `
[
  [🟦] [🟥] [🟦],
  [🟥] [🟦] [🟥],
  [🟥] [🟦] [🟦],
]
`;
